### DISCUSSION TITLE
Incorrect GPU auto-selection in certain environments

### DISCUSSION THREAD

**Discussion by taoizm** [talmolab/sleap#1315] [General]
Not sure if it's an issue that needs to be fixed, so I'll just post it here.

While using v1.3.0, I encountered a problem with GPU auto-selection that appears to fail in certain environments. In my setup, I have both a Quadro P620 for display and a GeForce RTX 2070 Super for computation. Despite correctly detecting the 2070 Super (which has 8192 MiB memory) on line 1975 of create_trainer_using_cli(), it actually ended up using the P620 (with 2048 MiB memory) instead.

Upon investigation, I discovered that while get_gpu_memory() retrieves a list of GPU memory sizes via nvidia-smi, get_all_gpus() called by use_gpu() fetches tf.config.list_physical_devices("GPU"), and the order of these two methods doesn't match, resulting in this issue.

To resolve this problem in my environment, I have set an environment variable CUDA_DEVICE_ORDER=PCI_BUS_ID to ensure a consistent GPU order.

I greatly appreciate your support and the convenience your product brings to my work!

**Comment by roomrys** [comment]
Hi @taoizm,

Thanks for investigating and reporting. This is an issue that needs to be fixed. We currently have an issue (and WIP pull request) for masked GPU selection via CUDA_VISIBLE_DEVICES. This sounds like something we can roll in while modifying the auto-GPU selection code. I will add this to the open issue.

Thanks,
Liezl
